third_party/move/move-compiler-v2/src/file_format_generator/peephole_optimizer/optimizers.rs

```diff
@@ -11,22 +11,20 @@ pub trait BasicBlockOptimizer {
     fn optimize(&self, block: &[Bytecode]) -> Vec<Bytecode>;
 }
 
-/// An optimizer for a fixed window of bytecode.
-/// The fixed window can be assumed to be within a basic block.
-pub trait FixedWindowOptimizer {
-    /// The fixed window size for this optimizer.
-    fn fixed_window_size(&self) -> usize;
-
-    /// Given a fixed `window` of bytecode of size `self.fixed_window_size()`,
-    /// optionally return its optimized version.
+/// An optimizer for a window of bytecode within a basic block.
+/// The window is always a suffix of a basic block.
+pub trait WindowOptimizer {
+    /// Given a `window` of bytecode, return a tuple containing:
+    ///   1. an optimized version of a non-empty prefix of the `window`.
+    ///   2. size of this prefix (should be non-zero).
     /// If `None` is returned, the `window` is not optimized.
-    fn optimize_fixed_window(&self, window: &[Bytecode]) -> Option<Vec<Bytecode>>;
+    fn optimize_window(&self, window: &[Bytecode]) -> Option<(Vec<Bytecode>, usize)>;
 }
 
-/// A processor to perform fixed window optimizations of a particular style on a basic block.
-pub struct FixedWindowProcessor<T: FixedWindowOptimizer>(T);
+/// A processor to perform window optimizations of a particular style on a basic block.
+pub struct WindowProcessor<T: WindowOptimizer>(T);
 
-impl<T: FixedWindowOptimizer> BasicBlockOptimizer for FixedWindowProcessor<T> {
+impl<T: WindowOptimizer> BasicBlockOptimizer for WindowProcessor<T> {
     fn optimize(&self, block: &[Bytecode]) -> Vec<Bytecode> {
         let mut old_block = block.to_vec();
         // Run single passes until code stops changing.
```

Comment:
It seems likely this can be optimized, but we can wait and see how slow it is first.

Reply:
SG. Right now, there does not seem to be any statistical difference between compilation with this on and off, when compiling the aptos framework and its dependencies.